src/gui/PianoView.cpp

```diff
@@ -502,7 +544,52 @@ void PianoView::mousePressEvent(QMouseEvent *me)
 	}
 }
 
+bool PianoView::setMarkerKeyValue(IntModel *noteModel, int key_num, bool ignoreConstraints)
+{
+	if (!ignoreConstraints)
+	{
+		if (m_piano->instrumentTrack()->firstKeyModel() == noteModel && !firstMarkerAllowed(key_num)) return false;
+		if (m_piano->instrumentTrack()->lastKeyModel() == noteModel && !lastMarkerAllowed(key_num)) return false;
+	}
+	noteModel->setValue(static_cast<float>(key_num));
+	if (noteModel == m_piano->instrumentTrack()->baseNoteModel()) { emit baseNoteChanged(); }	// TODO: not actually used by anything?
+	return true;
+}
 
+bool PianoView::firstMarkerAllowed(int key_num)
```

Comment:
Totally IMO, but I find these functions overkill

Reply:
It is because the functionality is used from two different places - for the context menu itself and also when setting the value with constraints.